### PR TITLE
Recreate changes for explicit flow order from #1261

### PR DESCRIPTION
**Pull request by default0**
This adds a separate order value that gets tracked alongside each child added to a FlowContainer effectively ensuring that the z-ordering of children has no influence over their position in a layout produced by a FlowContainer.

This change required fixing up TabControl which relied on the old behaviour and adjusted z-ordering of children in order to change their positions in a layout.
While reworking TabControl I also noticed that the Items property returned the list of values not in the displayed order, but in reversed display order, which I changed, since it conflicted with what the documentation said about the Items property.

Recreates #1261 due to demand in #1303

### CODE REVIEW

**Comment by smoogipoo** [comment]
Okay so there are two main things that make this very awkward to use:

1. We need a way to retrieve layout positions - a get method.
2. Changing layout position doesn't change the iteration order. If, for example, you want to reorder the elements within a FlowContainer through layout positions, the iteration order of the items will not change even though the ordering has changed visually.

**Comment by default0** [comment]
@smoogipoo would need some clarification on what exactly you need:
1) Do you want a "get" method that returns the index into the layout of a specific child? Or do you want to retrieve the ``float`` value set as the layout position? Specifically, what is your use-case for it?
2) By iteration order you mean the iteration order of the ``Children`` property? If you want to iterate the children in order of their layout position from a ``FlowContainer`` what is stopping you from using ``FlowingChildren`` instead?

**Comment by smoogipoo** [comment]
1. Layout position of a specific child. Super simple method.
2. Yes, of `Children`. Also of any LINQ extensions as part of the interfaces that Container implements, such as `First()` - which all use Children. FlowingChildren is protected in this case, and even if it were public, that's potentially a pretty expensive operation that may be done every frame (in this case, with every drag event).

Consider the use case of moving elements within the playlist and having a button to select the previous/next track in the list.

**Comment by default0** [comment]
That didn't particularly answer 1). I still don't know if you would prefer GetLayoutPosition() to return an ``int`` or a ``float``.
Thanks for the clarification otherwise :-)

**Comment by smoogipoo** [comment]
GetLayoutPosition -> return float.

**Comment by smoogipoo** [comment]
Moving this to Candidate Issues since there hasn't been any further work on this.

**Comment by default0** [comment]
Unsure there's a reasonable way to comply with your requirement of making Children have the same order as FlowChildren. The order of Children is what determines draw order, and the whole point is to have draw order and flow order be two distinct concepts (as they should be). It could probably be done by reimplementing / reusing certain parts of CompositeDrawable / CompositeDrawNode, but I feel like that blows this issue way out of proportion. So, if that requirement is non-negotiable and all your Linq or generally code that deals with container children relies on Children order being flow order, then I simply don't see a reason to have this change, useful as it is for certain scenarios in which FlowContainers are used (prime example still being dropdowns).

tl;dr if the Children requirement is not lifted and no one else is stepping up to tackle this, you might as well close this, as I am not going to be implementing nor want to be responsible for the resulting mess.

**Comment by smoogipoo** [comment]
I agree that draw order and flow order should be different concepts, but when you're looking at a flowcontainer and its children on the screen the ordering that's most apparent is the flow order.

If a separate iterator is possible to be utilised that gives you the flow-order sorted children, then I'd be content with that. Maybe repurpose `FlowingChildren` to represent all children, and create an `AliveFlowingChildren` to represent the alive children? That might be a good direction to take.

**Comment by default0** [comment]
I understand where you're coming from, but consider that the iteration order of Children also has implication for input handling, so if you have visual overlap in a flow container and the input logic of the framework would want to correctly give mouse input to the children of the flow container, it could **not** use the Children property for this anymore and would need to special case this (since Children would not be giving any indication about z-ordering for flow containers).

Having a second property by exposing FlowingChildren and figuring out the story with children being alive or dead around that sounds a lot more feasible and a lot less messy and thus a lot less prone to bizarre corner-cases (like the input one mentioned above). If that's the requirement going forward, I can probably get around to this somewhen this or next week.

**Comment by smoogipoo** [comment]
I would be fine with that - FlowingChildren/AliveFlowingChildren.

**Comment by default0** [comment]
``FlowingChildren`` is based off only alive internal children - and since this property is to be used by subclasses to determine positioning in a layout, there is really no concept of "flow order" for dead children. Therefore a distinction between ``FlowingChildren`` and a hypothetical ``AliveFlowingChildren`` is nonsensical based on the behaviour implemented in this PR.

I am also not sure if this distinction is at all useful, since ``FlowingChildren`` ought to be a representation of "the children that are being flowed, in order of their appearance in the flow", and "the children that are being flowed" never includes dead children.

If you have a specific use-case in mind that cannot be done properly with the way things are currently handled, please explain so I can understand where you require that distinction.

**Comment by swoolcock** [comment]
What's the status of this PR?  I would like to finish up #1363 but I'm waiting on a safer way to determine tab order.

**Comment by default0** [comment]
Waiting on feedback from ppy or smoogi to either revise or get approval & rebase on master.

**Comment by peppy** [comment]
Will run this through its paces now. If all looks well at osu! end I'm fine with this change.

**Comment by peppy** [comment]
I don't follow how this change should be implemented.

How should a reverse flow order be done now?

``csharp
// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE

using System.Collections.Generic;
using System.Linq;
using osu.Framework.Graphics;
using osu.Framework.Graphics.Containers;

namespace osu.Game.Graphics.Containers
{
    public class ReverseChildIDFillFlowContainer<T> : FillFlowContainer<T> where T : Drawable
    {
        protected override int Compare(Drawable x, Drawable y) => CompareReverseChildID(x, y);

        public override IEnumerable<Drawable> FlowingChildren => base.FlowingChildren.Reverse();
    }
}
```

**Comment by default0** [comment]
Your posted code works in a testcase, so I don't follow your confusion. Overriding FlowingChildren to change the order of flow is the intended way to implement different kinds of flow ordering.

Would you prefer a different approach to it?

**Comment by smoogipoo** [comment]
Everything looks to be in order osu!-side. Tested:
- Chat
- Playlist
- ButtonSystem (ofc)
- BeatmapSetOverlay
- BeatmapOptionsOverlay

I'm ready to try this out again.
